Add Early Cycle EnKF workflow

XianwuXue-NOAA opened the pull request:
**Description**
Updates workflow to allow early cycle EnKF to run.

Refs: #1021

**Type of change**
- [x] Bug fix (non-breaking change which fixes an issue)

**How Has This Been Tested?**
- [x] Cycled test on Hera
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code

XianwuXue-NOAA committed:
Fix a bug to exclude the unsued tasks for cycled gfs

 On branch feature/dev_early_enkf
	modified:   workflow/applications.py

Refs: #1021

WalterKolczynski-NOAA commented on workflow/applications.py:
```diff
@@ -409,9 +409,6 @@ def _get_cycled_task_names(self):
         if self.do_metp:
             gfs_tasks += ['metp']
 
-        if self.do_hybvar and 'gfs' in self.eupd_cdumps:
```
Is this really the correct solution? If these jobs shouldn't run for gfs, then I think they should be moved to the hybrid_gdas_tasks list instead of hybrid_gdas_or_gfs_tasks. But I'm not 100% sure we don't need these jobs for early cycle. @aerorahul?

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/dev_early_enkf

XianwuXue-NOAA committed:
Modify to add task for cycled gfs

 On branch feature/dev_early_enkf
	modified:   workflow/applications.py
	modified:   workflow/rocoto/workflow_tasks.py

Refs: #1021

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/dev_early_enkf

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -1075,7 +1075,7 @@ def atmensanalpost(self):
     def ecen(self):
 
-        self._is_this_a_gdas_task(self.cdump, 'ecen')
+        #self._is_this_a_gdas_task(self.cdump, 'ecen')
```
Just delete these instead of commenting them out.

XianwuXue-NOAA replied:
Thanks. Yes, I will delete them later.

XianwuXue-NOAA committed:
Remove echgres for cycled gfs

 On branch feature/dev_early_enkf
	modified:   workflow/applications.py

Refs: #1021

XianwuXue-NOAA committed:
Change eupd_cdump to cdump and change hard coded gdas to cdump
    for cycled gfs

 On branch feature/dev_early_enkf
	modified:   workflow/rocoto/workflow_tasks.py

Refs: #1021

XianwuXue-NOAA committed:
Change back for echgres funtion because cycled gfs doesn't run it

 On branch feature/dev_early_enkf
	modified:   rocoto/workflow_tasks.py

Refs: #1021

WalterKolczynski-NOAA commented:
@aerorahul Other than the commented out lines that need to be removed, is there anything holding this up?

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA  Still working on it. We may change it back (see the last comment), when everything is done, I will delete all commented lines.

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/dev_early_enkf

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/dev_early_enkf

XianwuXue-NOAA committed:
Improve to run early enkf

 On branch feature/dev_early_enkf
 Changes to be committed:
	modified:   jobs/JGDAS_ENKF_ECEN
	modified:   jobs/JGDAS_ENKF_SFC
	modified:   parm/config/config.anal
	modified:   ush/calcanl_gfs.py

Refs: #1021

XianwuXue-NOAA commented:
@aerorahul I have submitted my changes to this PR

XianwuXue-NOAA commented on ush/calcanl_gfs.py:
```diff
@@ -22,7 +22,8 @@ def calcanl_gfs(DoIAU, l4DEnsVar, Write4Danl, ComOut, APrefix, ASuffix,
     IAUHH = IAUHrs
     if Cdump == "gfs":
-        IAUHH = list(map(int,'6'))
+        #IAUHH = list(map(int,'6'))
```
I am not sure whether this change will affect others, so I keep the organize statement. After we make the final decision, I will clean this up.

WalterKolczynski-NOAA replied:
I think this may break IAU

WalterKolczynski-NOAA replied:
I guess this isn't needed, so it should be deleted.

XianwuXue-NOAA replied:
For this one, I think I can remove the "if" since else has the same statement with "if"

WalterKolczynski-NOAA commented on jobs/JGDAS_ENKF_ECEN:
```diff
@@ -80,7 +80,7 @@ fi
 export OPREFIX="${CDUMP}.t${cyc}z."
 export APREFIX="${CDUMP}.t${cyc}z."
 export APREFIX_ENKF="${CDUMP_ENKF}.t${cyc}z."
-export GPREFIX="${CDUMP}.t${gcyc}z."
```
Will this not mess up late-cycle?

XianwuXue-NOAA replied:
I am not sure, however, I think we can add some if statements to make it better for both early and late enkf

XianwuXue-NOAA replied:
Done. Thanks @WalterKolczynski-NOAA @TravisElless-NOAA

WalterKolczynski-NOAA commented on jobs/JGDAS_ENKF_ECEN:
```diff
@@ -96,7 +96,7 @@ fi
 export COMIN="$ROTDIR/$CDUMP.$PDY/$cyc/$COMPONENT"
 export COMIN_ENS="$ROTDIR/enkf$CDUMP_ENKF.$PDY/$cyc/$COMPONENT"
 export COMOUT_ENS="$ROTDIR/enkf$CDUMP.$PDY/$cyc/$COMPONENT"
-export COMIN_GES_ENS="$ROTDIR/enkf$CDUMP.$gPDY/$gcyc/$COMPONENT"
```
Same potential issue here.

WalterKolczynski-NOAA commented on jobs/JGDAS_ENKF_SFC:
```diff
@@ -97,7 +97,7 @@ export COMIN="$ROTDIR/$CDUMP.$PDY/$cyc/$COMPONENT"
 export COMIN_GES="$ROTDIR/$CDUMP.$gPDY/$gcyc/$COMPONENT"
 export COMIN_ENS="$ROTDIR/enkf$CDUMP_ENKF.$PDY/$cyc/$COMPONENT"
 export COMOUT_ENS="$ROTDIR/enkf$CDUMP.$PDY/$cyc/$COMPONENT"
-export COMIN_GES_ENS="$ROTDIR/enkf$CDUMP.$gPDY/$gcyc/$COMPONENT"
```
And here.

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -1104,15 +1104,15 @@ def _get_ecengroups():
             return grp, dep, lst
 
-        eupd_cdump = 'gdas' if 'gdas' in self.app_config.eupd_cdumps else 'gfs'
+        #eupd_cdump = 'gdas' if 'gdas' in self.app_config.eupd_cdumps else 'gfs'
```
Delete dead code.

XianwuXue-NOAA replied:
Just keep is as reference, and will delete dead code later.

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -1133,28 +1133,28 @@ def _get_ecengroups():
     def esfc(self):
 
-        self._is_this_a_gdas_task(self.cdump, 'esfc')
+        #self._is_this_a_gdas_task(self.cdump, 'esfc')
```
More dead code. Not going to list them all.

XianwuXue-NOAA replied:
Yes, just keeping them as ref and will delete them all later

WalterKolczynski-NOAA reviewed (changes requested):
I'm not sure these changes leave late cycle functionality intact.

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/dev_early_enkf

XianwuXue-NOAA commented:
@CatherineThomas-NOAA

- expdir on Hera: /scratch2/NCEPDEV/ensemble/noscrub/Xianwu.Xue/GEFS_v13_da/test_early_enkf_cold_BOTH_c96_test # my experiment is using C96/C48 with 20 members.
- Output: /scratch1/NCEPDEV/stmp2/Xianwu.Xue/o/test_early_enkf_cold_BOTH_c96_test
- ICs: /scratch2/NCEPDEV/ensemble/noscrub/Xianwu.Xue/GEFS_v13_da/z_WalterData/C96_cold_IC[C384_cold_IC]

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/dev_early_enkf

XianwuXue-NOAA committed:
Delete the unused statements

 On branch feature/dev_early_enkf
 Changes to be committed:
	modified:   workflow/rocoto/workflow_tasks.py

Refs: #1021

XianwuXue-NOAA committed:
Update to GDUMP to point the early-cycle to the correct directory
 Changes to be committed:
	modified:   jobs/JGDAS_ENKF_SFC

Refs: #1021

XianwuXue-NOAA committed:
Modify to run gfsefcsN for early cycle enkf

 Changes to be committed:
	modified:   jobs/JGDAS_ENKF_FCST
	modified:   ush/forecast_predet.sh

Refs: #1021

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/dev_early_enkf

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA @TravisElless-NOAA I have completed one and half tests successfully. The log files are in : /scratch1/NCEPDEV/stmp2/Xianwu.Xue/o/test_early_enkf_cold_BOTH_c96/logs/2022012112. I am working on solving the issues which caused failure in lint.

XianwuXue-NOAA committed:
To fix lint error

 On branch feature/dev_early_enkf
 Changes to be committed:
	modified:   ush/forecast_predet.sh

Refs: #1021

XianwuXue-NOAA committed:
To fix the lint error Error: SHELLCHECK_WARNING

Prefer [ p ] && [ q ] as [ p -a q ] is not well defined.

 On branch feature/dev_early_enkf
 Changes to be committed:
	modified:   ush/forecast_predet.sh

Refs: #1021

XianwuXue-NOAA commented:
I have fix the lint error.

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA @TravisElless-NOAA @CatherineThomas-NOAA @aerorahul Do you have any comments about this PR?  If you have, please let me know. I will merge the latest develop branch and then do the final test.

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/dev_early_enkf

WalterKolczynski-NOAA commented:
@XianwuXue-NOAA If the DA group has signed off on the results, the only thing I'm waiting on is removal of the dead code.

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA Thanks. I will remove them today after the meeting this afternoon.

XianwuXue-NOAA committed:
Remove the dead code

 On branch feature/dev_early_enkf
 Changes to be committed:
	modified:   workflow/rocoto/workflow_tasks.py

Refs: #1021

XianwuXue-NOAA committed:
Delete the unused codes

 On branch feature/dev_early_enkf
 Changes to be committed:
	modified:   ush/calcanl_gfs.py

Refs: #1021

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA I have merged the latest develop branch into my branch and also delete the dead codes. Travis will triple check the changes and results.

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/dev_early_enkf

WalterKolczynski-NOAA commented:
Waiting on @CatherineThomas-NOAA to give final approval before merging.

CatherineThomas-NOAA reviewed (approved):
Code looks good to me.  Thanks @XianwuXue-NOAA  and @TravisElless-NOAA for testing.  Approved.

XianwuXue-NOAA commented:
@TravisElless-NOAA @CatherineThomas-NOAA and @WalterKolczynski-NOAA Thanks a lot for your approval.
